lib/src/merged_tree.rs

```diff
@@ -975,14 +973,17 @@ impl<'matcher> TreeDiffStreamImpl<'matcher> {
     }
 
     fn poll_tree_futures(&mut self, cx: &mut Context<'_>) {
-        let mut pending_index = 0;
-        while pending_index < self.pending_trees.len()
-            && (pending_index < self.max_concurrent_reads
-                || self.items.len() < self.max_queued_items)
+        let mut current_dir: Bound<RepoPathBuf> = Bound::Unbounded;
+        let mut num_polled = 0;
+        while let Some((dir, future)) = self
+            .pending_trees
+            .range_mut((current_dir, Bound::Unbounded))
+            .next()
```

Comment:
nit: since this has to look up `current_dir` inside a loop, I think it's simpler to use plain `for _ in &mut ` loop and delete completed futures by later pass.